src/net_processing.cpp

```diff
@@ -3262,6 +3290,55 @@ bool PeerLogicValidation::SendMessages(CNode* pto, std::atomic<bool>& interruptM
             }
         }
 
+        // Check that outbound peers have reasonable chains
+        if (!(state.fProtectFromDisconnect || pto->fInbound || pto->m_manual_connection || pto->fFeeler || pto->fOneShot) && state.fSyncStarted) {
+            // This is an outbound peer subject to disconnection if their chain
+            // lags behind ours (note: if their chain has more work than ours,
+            // we should sync to it, unless it's invalid, in which case we
+            // should find that out and disconnect from them elsewhere).
+            if (state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= chainActive.Tip()->nChainWork) {
+                if (state.nHeadersChainTimeout != 0) {
+                    state.nHeadersChainTimeout = 0;
+                    state.header_with_required_work = nullptr;
+                    state.fSentGetHeadersToCheckChainSync = false;
+                }
+            } else if (state.nHeadersChainTimeout == 0 || (state.header_with_required_work != nullptr && state.pindexBestKnownBlock != nullptr && state.pindexBestKnownBlock->nChainWork >= state.header_with_required_work->nChainWork)) {
+                // Our best block known by this peer is behind our tip, and we're either noticing
+                // that for the first time, OR this peer was able to catch up to some earlier point
+                // where we checked against our tip.
+                // Either way, set a new timeout based on current tip.
+                state.nHeadersChainTimeout = nNow + CHAIN_SYNC_TIMEOUT;
```

Comment:
Should we only be setting this if `state.nHeadersChainTimeout` is unset or less than `nNow`? Otherwise it seems like we if we fall into this case due to the second half of the predicate (i.e. `nHeadersChainTimeoue != 0`), we might keep bumping the timeout each time `SendMessages` is called with this peer. If this is the case, we may never hit the timeout logic below.